src/ExpressionEvaluator/Core/Source/ExpressionCompiler/CustomTypeInfo.cs

```diff
@@ -1,37 +1,196 @@
+            ReadOnlyCollection<string> tupleElementNames)
+        {
+            var payload = Encode(dynamicFlags, tupleElementNames);
+            return (payload == null) ? null : DkmClrCustomTypeInfo.Create(PayloadTypeId, payload);
+        }
+
+        /// <summary>
+        /// Return a copy of the custom type info without tuple element names.
+        /// </summary>
+        internal static DkmClrCustomTypeInfo WithNoTupleElementNames(this DkmClrCustomTypeInfo typeInfo)
+        {
+            if ((typeInfo == null) || (typeInfo.Payload == null) || typeInfo.PayloadTypeId != PayloadTypeId)
+            {
+                return typeInfo;
+            }
+
+            var payload = typeInfo.Payload;
+            int length = payload[0] + 1;
+            if (length == payload.Count)
+            {
+                return typeInfo;
+            }
+
+            return DkmClrCustomTypeInfo.Create(PayloadTypeId, new ReadOnlyCollection<byte>(CopyBytes(payload, 0, length)));
+        }
+
+        /// <summary>
+        /// Return a copy of the custom type info with the leading dynamic flag removed.
+        /// There are no changes to tuple element names since this is used for walking
+        /// into an array element type only which does not affect tuple element names.
+        /// </summary>
+        internal static DkmClrCustomTypeInfo SkipOne(DkmClrCustomTypeInfo customInfo)
+        {
+            if (customInfo == null)
+            {
+                return customInfo;
+            }
+
+            ReadOnlyCollection<byte> dynamicFlags;
+            ReadOnlyCollection<string> tupleElementNames;
+            CustomTypeInfo.Decode(
+                customInfo.PayloadTypeId,
+                customInfo.Payload,
+                out dynamicFlags,
+                out tupleElementNames);
+
+            if (dynamicFlags == null)
+            {
+                return customInfo;
+            }
+
+            return Create(DynamicFlagsCustomTypeInfo.SkipOne(dynamicFlags), tupleElementNames);
+        }
+
+        // Encode in payload as a sequence of bytes {count}{dynamicFlags}{tupleNames}
+        // where {count} is a byte of the number of bytes in {dynamicFlags} (max: 8*256 bits)
+        // and {tupleNames} is a UTF8 encoded string of the names separated by '|'.
+        internal static ReadOnlyCollection<byte> Encode(
+            ReadOnlyCollection<byte> dynamicFlags,
+            ReadOnlyCollection<string> tupleElementNames)
+        {
+            if ((dynamicFlags == null) && (tupleElementNames == null))
+            {
+                return null;
+            }
+            var builder = ArrayBuilder<byte>.GetInstance();
+            if (dynamicFlags == null)
+            {
+                builder.Add(0);
+            }
+            else
+            {
+                int length = dynamicFlags.Count;
+                if (length > byte.MaxValue)
+                {
+                    // TODO: Test.
```

Comment:
Yes, sorry, there are several TODOs in this change. I will address the issues or log bugs before closing the PR.